### PR TITLE
added gpu support for rpa

### PR DESCRIPTION
**Pull request by JensWehner**
I added a bit of gpu support for the rpa evaluation. 

@baumeier can you do a quick benchmark with the complex integration?

### CODE REVIEW

**Comment by baumeier** [comment]
```
... ... 2020-11-22 18:6:22 Calculated Hartree exchange contribution
... ... 2020-11-22 18:6:22 Scissor shifting DFT energies by: 0 Hrtterminate called after throwing an instance of 'std::runtime_error'
  what():  Shape mismatch in cuda diag_gemm
```

**Comment by JensWehner** [comment]
thx so more tests are needed. :/

**Comment by JensWehner** [comment]
@baumeier can you try again?

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by JensWehner** [comment]
@votca-bot changelog: Added GPU support for RPA

**Comment by baumeier** [comment]
```
 63%] Building CXX object xtp/src/libxtp/CMakeFiles/votca_xtp.dir/cudapipeline.cc.o
remark #11074: Inlining inhibited by limit max-size
remark #11076: To get full report use -qopt-report=4 -qopt-report-phase ipo
/home/bbaumeie/votca_sources/votca/xtp/src/libxtp/openmp_cuda.cc(153): error: no operator "+=" matches these operands
            operand types are: Eigen::MatrixXd += votca::xtp::CudaMatrix
    reduction_[0] += *D;
                  ^

compilation aborted for /home/bbaumeie/votca_sources/votca/xtp/src/libxtp/openmp_cuda.cc (code 2)
```

**Comment by JensWehner** [comment]
last fix I hope.

**Comment by baumeier** [comment]
Now it is not complaining, at least. Is super slow though.

**Comment by baumeier** [comment]
How is the balance here between OpenMP and CUDA?

**Comment by JensWehner** [comment]
what hmmm

**Comment by baumeier** [comment]
Single thread CPU: 32 sec
Single thread CPU + GPU: 103 sec

**Comment by felipeZ** [comment]
@JensWehner
1. Have you compare the performance of No GPU, 1 GPU vs 2 GPUs?
2. It would be great if you run [Nvidia Nsight](https://developer.nvidia.com/nsight-systems) to profile your implementation. You probably should it an small system because the profiling takes some time

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by JensWehner** [comment]
@felipeZ  Actually the transfers for reasonably large matrices are not that bad at all

Screenshot_2020-12-02_00-05-20

**Comment by felipeZ** [comment]
@JensWehner it seems that the GPU are doing some work apart from copy things back and for, good job!
I also thing that using multiple `streams` can significantly speed up the calculations

**Comment by JensWehner** [comment]
actually it is like 80% work. The 3c fill stuff is only at 80% I can show that later.

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by JensWehner** [comment]
@baumeier it should work now. Do you want to give it a try. :)

**Comment by baumeier** [comment]
Will do sometime later tomorrow. Does this now work with the CDA?

**Comment by JensWehner** [comment]
yes, so the GPUs support the cda computation as well, but it is only parallelized via the qp levels.

**Comment by baumeier** [comment]
@JensWehner only using a single GPU, it seems

**Review by baumeier — APPROVED** [review]

**Comment by junghans** [comment]
@votca-bot copyright

**Comment by JensWehner** [comment]
please do not merge yet, waiting for @felipeZ review. @rubengerritsen could you also have a look and see if you understand what is happening otherwise I will refactor again.

**Review by felipeZ — APPROVED** [review]
@JensWehner  great job!
I still think that adding some comments on the most complex CUDA/OpenMP interactions will help some newcomers

**Comment by rubengerritsen** [comment]
I can't comment on non edited parts for some reason, so here with a permalink https://github.com/votca/xtp/blob/2c483cf2c8cbfe2dab185956d77f9aa4ff74b760/src/libxtp/cudamatrix.cc#L35 
this seems to always return `Index(count)`, it won't do anything with the error

**Review by rubengerritsen — APPROVED** [review]
Looks really nice. @JensWehner It is okay to understand, I needed a bit of help from the cuda documentation and the BLAS wikipedia page to get started, but once I glanced over those everything was very understandable.

**Comment by JensWehner** [comment]
> I can't comment on non edited parts for some reason, so here with a permalink
> 
> https://github.com/votca/xtp/blob/2c483cf2c8cbfe2dab185956d77f9aa4ff74b760/src/libxtp/cudamatrix.cc#L35
> 
> this seems to always return `Index(count)`, it won't do anything with the error

good catch actually.

File: src/libxtp/openmp_cuda.cc [hunk 1]
```diff
@@ -23,16 +23,42 @@
 namespace votca {
 namespace xtp {
 
+OpenMP_CUDA::OpenMP_CUDA() {
```
[the remark this answers is not in the document]
Comment (JensWehner) [reply src/libxtp/openmp_cuda.cc:26]: Heyho I added a description in the header file.

**Comment by JensWehner** [comment]
@votca-bot format

**Pull request merged**